### PR TITLE
[nomerge][2.12] Welcome back Partest!

### PR DESCRIPTION
**Pull request by dwijnand**
The motivation is to bring "/* scalac: ... */" pragma support to 2.12.x's partest, as easily as possible, as backporting/forward-merging changes is frequently impacted by 2.12's lack of pragma support and 2.13's lack of flags-file support...

So, as an alternative to scala/scala-partest#118, here's a without-history c+p of 2.13.x's partest, for the 2.12.x branch.

I decided to defer backporting the "testkit" split out and other related build tweaks that followed partest's re-integration into the 2.13.x branch (though I might've copied some details), particularly I had been looking at:

* https://github.com/scala/scala/pull/7817
* https://github.com/scala/scala/pull/7847
* https://github.com/scala/scala/pull/8378

I also decided to defer both migrating all the flags files to pragmas as well as dropping flags support (which I actually had to restore - all the other post-copy tweaks were just mechanical changes, like scala.util.chaining and JavaConverters.)

I've tried to make a reasonable attempt at putting together a workable change, without sinking too much time into this.

But feedback welcome, particular by @som-snytt.  Specifically could you check the `ExtConsoleReporter`/`DirectCompiler` close/finish/hasError logic, please?

### CODE REVIEW

File: src/partest/scala/tools/partest/nest/Runner.scala [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+
+  /** Source files for the given test file. */
+  def sources(file: File): List[File] = if (file.isDirectory) file.listFiles.toList.filter(_.isJavaOrScala) else List(file)
+
+  def newCompiler = new DirectCompiler(this)
+
+  def attemptCompile(sources: List[File]): TestState = {
+    // TODO: Deferred until 2.12.x is ready to lose its flags files, & possibly flags file support entirely
+    //val badflags = (testFile :: (if (testFile.isDirectory) sources else Nil)).map(_.changeExtension("flags")).find(_.exists)
+    //if (badflags.isDefined) genFail(s"unexpected flags file ${badflags.get} (use source comment // scalac: -Wfatal-warnings)")
```
Comment (dwijnand) [inline src/partest/scala/tools/partest/nest/Runner.scala]: 🚏

File: src/partest/scala/tools/partest/nest/DirectCompiler.scala [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    def execCompile() =
+      if (command.shouldStopWithInfo) {
+        logWriter append (command getInfoMessage global)
+        runner genFail "compilation stopped with info"
+      } else {
+        new global.Run compile sources.map(_.getPath)
+        if (!reporter.hasErrors) runner.genPass()
+        else {
+          reporter.printSummary()
+          reporter.writer.close()
```
Comment (dwijnand) [inline src/partest/scala/tools/partest/nest/DirectCompiler.scala:136]: ❓

**Comment by dwijnand** [comment]
```
 
!! 1243 - pos/t8187.scala                           [unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)]
ok 1244 - pos/t8177d.scala                        
!! 1245 - pos/t8209a                                [unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)]
!! 1246 - pos/t8209b                                [unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)]
ok 1247 - pos/t8177b.scala                        
....
ok 1268 - pos/t8301b.scala                        
!! 1269 - pos/t8352                                 [unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)]
!! 1270 - pos/t8364.scala                           [unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)]
ok 1271 - pos/t8301.scala                         
!! 1272 - pos/t8369a.scala                          [unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)]
!! 1273 - pos/t8369b.scala                          [unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)
```

File: src/partest/scala/tools/partest/nest/Runner.scala [hunk 3]
```diff
@@ -0,0 +1,770 @@
...
+    val round1                  = if (scalaFiles.isEmpty) None else Some(ScalaAndJava(allFiles))
+    val round2                  = if (javaFiles.isEmpty) None else Some(OnlyJava(javaFiles))
+
+    List(round1, round2).flatten
+  }
+
+  def runPosTest(): TestState =
+    // TODO: Deferred until cleanup has happened
+    //if (checkFile.exists) genFail("unexpected check file for pos test (use -Xfatal-warnings with neg test to verify warnings)")
+    //else runTestCommon()
```
Comment (dwijnand) [inline src/partest/scala/tools/partest/nest/Runner.scala]: 😴

**Comment by som-snytt** [comment]
It's quite some pandemic busy work. I'm surprised there is any forward merging to 2.13, and I would be surprised if partest maintenance is worth the efforts. OTOH, repairing the Vulpix split might make maintenance easier for 2.13/3.0. I have tinkered with dotty but I've forgotten again how to run a test.

**Comment by SethTisue** [comment]
@sjrd will this affect Scala.js?

**Comment by sjrd** [comment]
It probably will, yes. But we're used to juggling with different versions of partest depending on the version of Scala, so nothing disruptive I think.

**Comment by som-snytt** [comment]
I might have to comment on [the youtube video](https://www.youtube.com/watch?v=Mmm3KTa601s), Anyone else here because of Welcome Back, Partest?

By coincidence, recently I've punned on Kotter when working in Kotlin.

**Review by lrytz — APPROVED** [review]
👍

> I'm surprised there is any forward merging to 2.13

We got bitten a few times by the mismatch. At least flags are not silently ignored in 2.13, but for backports the situation is still unsafe (`/* scalac: -Werror */` is silently ignored). And migrating between the two formats is really annoying.

**Comment by som-snytt** [comment]
I'm surprised there is any backporting to 2.12, but I can see how that would be annoying.

I'll look into a Scalafix for partests!

**Review by retronym — APPROVED** [review]
I'm happy to land this as is, but as discussed lets follow up with a migration of .flags files to pragmas and removal of support of .flags files.

**Comment by som-snytt** [comment]
If retronym is happy, I'm happy.

Another option so to speak is `partest -migrate`. It's tricky to catch that `-Werror` for a `pos` test matters. The ultimate would be a rig that tests both before and after, or before and HEAD, for fail and pass respectively.

Relatedly, I've wondered when it's preferable to piggy-back on an existing test vs adding a new test file. `t555` should fail before its associated fix. An edit to `t555` should also fail before the associated fix. But it's not reasonable to verify it manually.

**Comment by SethTisue** [comment]
proposal to archive the old repo: https://github.com/scala/scala-partest/pull/119

**Pull request merged**